### ISSUE TITLE
[BUG] Linking errors for __bitsScanReverse, __BitScanForward in 64 bit

### ISSUE THREAD

**Issue by bhupeshpant19jan** [android-ndk/ndk#949]
#### On 64 bit, I am getting linking errors for MS intrinsics functions like __bitScanForward, __bitScanReverse and other functions.

`
nning: f:\nugetcache\androidndk.16.1.10\toolchains\llvm\prebuilt\windows-x86_64\bin\clang++.exe @e:\office\build\droidarm64\debug\xxx\warp\warp10\linkdll\objd\droidarm64\linker.rsp -o E:\Office\Target\droidarm64\debug\xxx\x-none\unstripped\somrandom.so
e:/office/dev/xxx/yyy/Libraries/Shared/public/testquad.h:216: undefined reference to '_BitScanReverse64'
E:\Office\dev\xxx\warp\warp10\umdriver/../Inc/sdsds.h:1817: undefined reference to '_byteswap_ulong'
E:\Office\dev\xxx\warp\FormatLib.h:1811: undefined reference to '_byteswap_ulong'
E:\Office\dev\xxx\jit/../Inc/AlphaBlt.h:4547: undefined reference to '_byteswap_ulong'
E:\Office\dev\directx\warp\warp10\jit/../Inc/AlphaBlt.h:4671: undefined reference to '_byteswap_ulong'
e:\office\dev\xxx\sdfd.cpp:44: undefined reference to '__umulh'
Exit code: 1
`

The above code is working fine for x86 build. I have also tried to replicate the issue locally in my test app. I have first build it on x86 build in which case it builds fine but when I am trying to build it in x64, it is causing similar linking issues.
It's clear that in 64bit  Clang doesn't provide the implementation for these intrinsic.

#### Environment Details

Not all of these will be relevant to every bug, but please provide as much
information as you can.

* NDK Version: r16
* Build system: ndk-build? CMake? Standalone toolchain? Custom? (If gradle, file
  a bug against Android Studio as specified above.) Nmake
* Host OS: Windows, Mac, or Linux? Which Linux? Windows
* ABI: 
* NDK API level: 23
* Device API level:

**Comment by enh** [comment]
as a workaround for the byte swaps, you can use `__builtin_bswap16`/`__builtin_bswap32`/`__builtin_bswap64`. if you search your NDK for `intrin.h` you'll see workarounds for all these.

but @stephenhines will know whether these MSVC extensions are supposed to work anyway (or only work if you're targeting Windows, or ...).

**Comment by stephenhines** [comment]
This isn't actionable without a repro case. Since none of these symbols are generated by Clang itself, I strongly suspect that they are for building Windows objects (which is not a supported use case of the Android NDK). If you can send me something that reproduces the bug, I can look further at why these calls are present, but I suspect that it is due to the sources themselves calling Windows-specific functionality.

**Comment by bhupeshpant19jan** [comment]
Yes, we are trying to call the cross-plat code that contains this functionality. But my question is if these are provided in x86 build why not it is being provided by x64. In x86 we are able to successfully compile with these MS specific dependencies.

**Comment by bhupeshpant19jan** [comment]
I am able to replicate the scenario using a local test app.

**Comment by stephenhines** [comment]
I am asking for a standalone piece of sample code to reproduce the issue. As it is, this looks like you are trying to target a Windows host binary with the NDK, which is not a supported use-case for our toolchain (mostly because we don't supply the necessary headers/libraries). Can you attach code + the full command line? Note that the command line above uses a .rsp file that has most of the arguments in it. Without seeing that, we don't know how to reproduce the problem you are having at all.

**Comment by bhupeshpant19jan** [comment]
Attaching test app projects for the reference.

**Comment by bhupeshpant19jan** [comment]
[64bittest.zip](https://github.com/android-ndk/ndk/files/3053431/64bittest.zip)

**Comment by stephenhines** [comment]
I finally managed to narrow down the problem, and it is actually in your source file. It has to do with the pragma/declarations that you are using. You still are using `long *Index` when it needs to be `int *Index` to work. The reason for this is that `long` expands to a 32-bit integer type on all versions of Windows. Since this is a Windows builtin, the code expects to see a 32-bit integer pointer type here. When it sees a `long`, which is 64-bit on AArch64 Android, it now sees this as an externally available function call. Then the toolchain doesn't lower it to the builtin, and produces just a call to `_BitScanReverse64()` with the 64-bit type signature.

I redeclared `_BitScanReverse64` using an `unsigned int *` instead, and everything worked. I'm not convinced that this is really a great solution, and think that it would probably be better to not depend on this builtin at all (or just reimplement it manually in your library).
